src/neuroglancer/sliceview/fpzip/fpzip_wasm.cpp

```diff
@@ -0,0 +1,255 @@
+/**
+ * @license
+ * Copyright 2022 William Silvermsith
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fpzip.h"
+#include <memory>
+
+class Fpzip {
+public:
+  unsigned int type;
+  unsigned int prec;
+  size_t nx;
+  size_t ny;
+  size_t nz;
+  size_t nf;
+
+  Fpzip () {
+		type = 0;
+		prec = 0;
+		nx = 0;
+		ny = 0;
+		nz = 0;
+		nf = 0;
+  }
+
+  Fpzip (unsigned char* buf) {
+		type = 0;
+		prec = 0;
+		nx = 0;
+		ny = 0;
+		nz = 0;
+		nf = 0;
+
+		decode_headers(buf);
+  }
+
+  Fpzip(Fpzip &orig) {
+		type = orig.type;
+		prec = orig.prec;
+		nx = orig.nx;
+		ny = orig.ny;
+		nz = orig.nz;
+		nf = orig.nf;
+  }
+
+  ~Fpzip() {
+  
+  }
+
+  size_t nvoxels() {
+		return nx * ny * nz * nf;
+  }
+
+  size_t nbytes() {
+		// type: 0 = float, 1 = double
+		return nvoxels() * (type + 1) * sizeof(float);
+  }
+
+  size_t get_type() { return type; }
+  size_t get_prec() { return prec; }
+  size_t get_nx() { return nx; }
+  size_t get_ny() { return ny; }
+  size_t get_nz() { return nz; }
+  size_t get_nf() { return nf; }
+
+  void decode_headers(unsigned char *data) {
+		FPZ* fpz = fpzip_read_from_buffer(static_cast<void*>(data));
```

Comment:
This seems to be a strange/broken API provided by fpzip, since it accepts `data` but does not know the number of bytes available.

Please investigate what the correct bounds check is and add it.

Reply:
I checked the header file and the missing num_bytes appears to be correct. https://github.com/LLNL/fpzip/blob/develop/include/fpzip.h#L192-L196

Reading the code, it seems the minimum byte stream (for the header) seems to be at least 6 x uint32_t.

Reply:
Previously, I found though that a minimum of 28 bytes are required, though I am not sure of why the last 4 are. I'll add that as a check.

Reply:
Actually I see that the actual decoding even after reading the header also does no bounds checking.  I don't think there is any safe way to use this API as is.  The upstream repository needs to be fixed to include proper bounds checking.

Reply:
Okay, I'm addressing this with the upstream maintainer. Will revisit this PR probably in a few days or weeks.